src/compiler.ts

```diff
@@ -10067,7 +10067,11 @@ export class Compiler extends DiagnosticEmitter {
 
   /** Evaluates a boolean condition, determining whether it is TRUE, FALSE or UNKNOWN. */
   evaluateCondition(expr: ExpressionRef): ConditionKind {
-    assert(getExpressionType(expr) == TypeRef.I32);
+    let type = getExpressionType(expr);
+    if (type == TypeRef.Unreachable)
+      return ConditionKind.UNKNOWN; // invalid expression given
```

Comment:
Not necessarily an "invalid" expression, but one that never returns (result is unreachable) according to Binaryen internals. The fix looks good, but I think the comment can be removed.